src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs

```diff
@@ -11,12 +12,15 @@ public class ProcessFetch
 	{
 		public void Process(FetchRequestBase resultOperator, QueryModelVisitor queryModelVisitor, IntermediateHqlTree tree)
 		{
-			var querySource = QuerySourceLocator.FindQuerySource(
```

Comment:
`QuerySourceLocator.FindQuerySource` can't reliably detect query source. Instead of actually query source it finds first source with "compatible" type